processor/transformprocessor/internal/common/functions.go

```diff
@@ -16,25 +16,40 @@ package common // import "github.com/open-telemetry/opentelemetry-collector-cont
 
 import (
 	"fmt"
-	"reflect"
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
 )
 
-var registry = map[string]interface{}{
-	"keep_keys":    keepKeys,
+type TransformFunction func(arguments []Value, pathParser PathExpressionParser, functions map[string]TransformFunction) (ExprFunc, error)
+
+var registry = map[string]TransformFunction{
 	"set":          set,
+	"keep_keys":    keepKeys,
 	"truncate_all": truncateAll,
 	"limit":        limit,
 }
 
 type PathExpressionParser func(*Path) (GetSetter, error)
 
-func DefaultFunctions() map[string]interface{} {
+func DefaultFunctions() map[string]TransformFunction {
 	return registry
 }
 
-func set(target Setter, value Getter) (ExprFunc, error) {
+func set(arguments []Value, pathParser PathExpressionParser, functions map[string]TransformFunction) (ExprFunc, error) {
```

Comment:
This makes defining a function much harder - we have a goal to make it easy to define functions as they will be common. Also ideally we eventually have an extension mechanism for adding functions in custom distros, so the function definition "API" needs to be easy to use for anyone.

I find reflective resolution of the arguments to be quite easy, but if avoiding reflection the scheme I would take is a `Config` type.

`func set(config FunctionConfig) {
  target, err := config.Setter(0)
  if err != nil return err
  value, err := config.Getter(1)
```

or `config.String(0)`, `config.Int(0)`, `config.GetSetter`, `config.Strings()`. etc.

While notably the error handling is more tedious, this is still reasonable I think. We definitely don't want to require functions to know about implementation tricks like `PathExpressionParser`

Reply:
I have some thoughts on how I landed on this approach, but I'm also not tied to it.  Wanted to throw it out here for discussion.  I'll make this PR a draft and lets continue the discussion in issue #10101.